design-proposals/generalize-virt-stack/gen-virt-stack.md

```diff
@@ -0,0 +1,140 @@
+- [LXC - Linux Containers](https://libvirt.org/drvlxc.html)
+- [OpenVZ](https://libvirt.org/drvopenvz.html)
+- [QEMU/KVM/HVF](https://libvirt.org/drvqemu.html)
+- [VirtualBox](https://libvirt.org/drvvbox.html)
+- [VMware ESX](https://libvirt.org/drvesx.html)
+- [VMware Workstation/Player](https://libvirt.org/drvvmware.html)
+- [Xen](https://libvirt.org/drvxen.html)
+- [Microsoft Hyper-V](https://libvirt.org/drvhyperv.html)
+- [Virtuozzo](https://libvirt.org/drvvirtuozzo.html)
+- [Bhyve - The BSD Hypervisor](https://libvirt.org/drvbhyve.html)
+- [Cloud Hypervisor](https://libvirt.org/drvch.html)
+
+There are several parts in the KubeVirt source-code that hard-code the use of the QEMU/KVM hypervisor driver, which prevents the creation of VMIs using another hypervisor driver. Therefore, KubeVirt needs to be updated to introduce the choice of the backend libvirt hypervisor driver to use for creating a given VMI. This would expand the set of scenarios in which KubeVirt can be used.
+
+## Goals
+
+KubeVirt should be able to offer a choice to its users over which libvirt hypervisor-driver they want to use to create their VMI. The aim is to remove any hard-coded dependency of KubeVirt on qemu-kvm and have it interact solely with libvirt, while still being able to provide the same functionality for qemu-kvm based VMIs.
+
+Additionally, this design proposal intends to provide developers of other hypervisors with a flexible way of onboarding their own hypervisor to KubeVirt.
+
+## Non Goals
+
+Support all libvirt hypervisor-drivers in KubeVirt. As the developers of hypervisor `H` make its libvirt hypervisor-driver stable, they would add functionality to KubeVirt so that it can leverage hypervisor `H` to create VMIs.
+
+## Definition of Users
+
+This proposal is aimed at serving users who intend to use KubeVirt on a cluster where at least one node has a virtualization stack different from the default QEMU-KVM stack.
+
+## User Stories
+
+- A user trying to use KubeVirt on a cluster of machines that have a hypervisor-VMM pair that is not necessarily QEMU/KVM.
+- A regular user of libvirt with any of its supported hypervisor drivers who now wants to leverage the orchestration capability provided by KubeVirt. But the user does not want to abandon their hypervisor driver of choice.
+- A regular user of a hypervisor-VMM-specific orchestration capability to expand their use to a cluster with a diverse set of hypervisor-VMM pairs.
+
+## Repos
+
+- KubeVirt
+- Libvirt
+
+# Design
+
+## API Changes
+
+Addition of a `vmi.hypervisor` field. Example of how a user could request a specific hypervisor as the underlying libvirt hypervisor-driver through the VMI spec:
+
+```yaml
+spec:
+  hypervisor: cloud-hypervisor
```

Comment:
I think here it is a good fit for something similar to the [kubernetes runtime classes](https://kubernetes.io/docs/concepts/containers/runtime-class/).  If we need additional configuration specific to the hypervisor they could go into a new CRD